### PR TITLE
fix(chat): prevent PEP 525 RuntimeError on stream disconnect and fix streaming log persistence

### LINKED ISSUE #49: [BUG] RuntimeError: async generator ignored GeneratorExit in streaming chat completions on client disconnect
## Description
When a client disconnects mid-flight during a streaming completion (e.g., closing browser tab, HTTP request timeout, or `curl` cancellation), the server throws an unhandled `RuntimeError: async generator ignored GeneratorExit`.

## Affected Location
- **File**: [app/routes/chat.py](file:///run/media/sourav/New%20Volume/OpenSource/OrcaRouter-Lite/app/routes/chat.py#L631-L650)
- **Lines**: 631–650 (inside `sse()` async generator `finally` block)

## Root Cause
In Python ([PEP 525](https://peps.python.org/pep-0525/)), an async generator processing `GeneratorExit` or `CancelledError` during teardown is strictly forbidden from yielding new values. 

In `app/routes/chat.py`, `yield "data: [DONE]\n\n"` is placed inside a `finally:` block:

```python
finally:
    try:
        yield "data: [DONE]\n\n"  # ❌ Fails when processing GeneratorExit / CancelledError
    except (asyncio.CancelledError, GeneratorExit):
        pass
```

When a client disconnects, `aclose()` or task cancellation throws `GeneratorExit` into the generator. The `finally:` block executes `yield`, causing Python's generator mechanism to raise:
`RuntimeError: async generator ignored GeneratorExit`

## Steps to Reproduce
1. Start an SSE stream via `POST /v1/chat/completions` with `"stream": true`.
2. Abort the HTTP connection mid-stream while tokens are printing.
3. Observe server logs or pytest warning:

```text
PytestUnraisableExceptionWarning: Exception ignored while closing generator 
<async_generator object chat_completions.<locals>.sse>: RuntimeError: async generator ignored GeneratorExit
```

## Reproduction Test Case (`pytest`)
```python
@pytest.mark.asyncio
async def test_sse_generator_ignored_generatorexit(monkeypatch):
    async def mock_stream():
        yield {"choices": [{"delta": {"content": "hello"}}]}
        await asyncio.sleep(10)
        yield {"choices": [{"delta": {"content": "world"}}]}

    fake_client = AsyncMock()
    fake_client.acompletion = AsyncMock(return_value=mock_stream())
    monkeypatch.setattr(router_cache, "get_router", AsyncMock(return_value=fake_client))

    response = await chat_completions(
        body=ChatCompletionRequest(model="gpt-4o-mini", messages=[{"role": "user", "content": "hi"}], stream=True),
        request=AsyncMock(),
        kc=KeyContext(key_id="test", workspace_id="default", name="test", key_type="standard"),
        db=AsyncMock(),
    )

    gen = response.body_iterator
    await gen.__anext__()  # read 1st chunk

    # Simulate client disconnect
    with pytest.raises(RuntimeError) as exc_info:
        await gen.aclose()
    
    assert "async generator ignored GeneratorExit" in str(exc_info.value)
```

## Expected Behavior
The async generator should terminate cleanly without raising unhandled `RuntimeError` server logs or corrupting task cancellation state.

## Proposed Fix
Move `yield "data: [DONE]\n\n"` out of the `finally:` block into the normal completion path of the `try:` block:

```diff
--- a/app/routes/chat.py
+++ b/app/routes/chat.py
@@ -553,6 +553,7 @@ async def sse() -> AsyncGenerator[str, None]:
                         agg_model = d["model"]
                     yield f"data: {json.dumps(d, separators=(',', ':'))}\n\n"
+            yield "data: [DONE]\n\n"
             except (asyncio.CancelledError, GeneratorExit):
                 ...
             finally:
-                try:
-                    yield "data: [DONE]\n\n"
-                except (asyncio.CancelledError, GeneratorExit):
-                    pass
                 with anyio.CancelScope(shield=True):
                     try:
                         await _finalize()
```

### PR DESCRIPTION
**Pull request by JhaSourav07**
Fixes #49 

## Description

This PR fixes two critical issues in the SSE streaming completions endpoint (`POST /v1/chat/completions`):

1. **PEP 525 Violation (`RuntimeError: async generator ignored GeneratorExit`)**:
   - **Problem**: `yield "data: [DONE]\n\n"` was executed inside the `finally:` block of the `sse()` generator. When a client disconnects mid-stream (or when task cancellation occurs), Starlette/AnyIO closes the generator by throwing `GeneratorExit`. Per Python PEP 525, yielding a value while handling `GeneratorExit` raises `RuntimeError: async generator ignored GeneratorExit`.
   - **Fix**: Moved `yield "data: [DONE]\n\n"` out of `finally:` to the normal end of the `try:` block execution. The `finally:` block now contains strictly non-yielding cleanup logic inside a shielded cancellation scope.

2. **Streaming Request Log Persistence Teardown Race**:
   - **Problem**: FastAPI closes the request-scoped `db` dependency context as soon as `chat_completions()` returns `StreamingResponse`. When `sse()` completes or disconnects later, `_finalize()` attempted `db.add(log)` and `await db.commit()` on a closed/detached session, causing 100% of streaming request logs to fail with `request_log_commit_failed`.
   - **Fix**: Updated `_finalize()` to acquire a fresh, isolated session from `session_mod._session_factory()` so log persistence succeeds reliably regardless of request context state.

---

## Type of Change

- [x] 🐛 **Bug fix** (non-breaking change fixing an issue)
- [x] 🧪 **Test update** (added test cases for stream cancellation/teardown)

---

### Automated Tests
1. **Targeted Disconnect Integration Test**:
   ```bash
   pytest tests/integration/test_stream_disconnect.py -s
   ```
   *Result*: `1 passed in 1.93s` — verifies `aclose()` closes cleanly without `RuntimeError`.

2. **Full Integration & Unit Test Suite**:
   ```bash
   pytest
   ```
   *Result*: `303 passed in 6.97s`

### Test Environment
- **OS**: Linux (Fedora / Ubuntu)
- **Python**: 3.14.6
- **Frameworks**: FastAPI / Starlette / AnyIO / SQLAlchemy Async Engine

---

## Checklist

- [x] My code follows the code style guidelines of this project.
- [x] I have performed a self-review of my code.
- [x] I have added a new unit/integration test (`tests/integration/test_stream_disconnect.py`).
- [x] New and existing tests pass locally with my changes.
- [x] No breaking API contract changes introduced.

### CODE REVIEW

**Comment by JhaSourav07** [comment]
Hi @fengyat @yi-here 

I've opened this PR to fix the unhandled RuntimeError: async generator ignored GeneratorExit during SSE streaming.

Everything is passing locally, including the new unit test reproducing the edge case. Whenever you have a chance, please take a look and let me know if any adjustments are needed!

Screenshot From 2026-08-04 17-59-54

**Comment by zhenjunchen-png** [comment]
Thanks for the detailed issue and the fix. Both changes look correct: moving the `[DONE]` sentinel out of the `finally:` block is the right fix for the PEP 525 violation, and using a fresh session in `_finalize()` correctly addresses the request-scoped session being closed before the stream finishes.

One request before merging:

**Emit `[DONE]` on the error path as well.** Previously the `finally:` block sent `[DONE]` after the error body. With this change, the `except Exception` branch yields the error payload but the stream ends without a `[DONE]` sentinel. Since that branch is not handling `GeneratorExit`, yielding there is safe:

```python
yield f"data: {json.dumps(err_body)}\n\n"
yield "data: [DONE]\n\n"
```

This keeps the behavior consistent for clients that read until `[DONE]`.

Also, CI is failing on lint: ruff reports `I001` (import block un-sorted) in `tests/integration/test_stream_disconnect.py`. Running `ruff check --fix .` locally and pushing should clear it. The pytest job only runs after lint passes.

Minor, non-blocking: `_finalize()` now reaches into `session_mod._session_factory` directly and duplicates the add/commit logic in two branches. It works, but if you want to tidy it up, a small helper in `packages/db/session.py` (or reusing `get_session()`) would avoid touching the private attribute. Fine to leave as is.

**Comment by zhenjunchen-png** [comment]
Merged. Thanks for the high-quality report and fix. Note: we pushed two small follow-up commits to your branch before merging: import sorting for the ruff gate, and emitting the [DONE] sentinel on the error path as discussed above. The unrelated CI failure in test_cache_group_check.py was fixed on main in #61. Looking forward to more contributions.
